src/api/tap.rs

```diff
@@ -0,0 +1,322 @@
+                                    );
+                                    sinks.insert(name.to_string(), id);
+                                } else {
+                                    error!(
+                                        message = "Couldn't connect component.",
+                                        component_name = ?name, id = ?id
+                                    );
+                                }
+
+                                Some(matched)
+                            }
+                            _ => {
+                                debug!(
+                                    message="Component not matched.",
+                                    component_name = ?name, patterns = ?patterns
+                                );
+                                None
+                            },
+                        }
+                    })
+                    .flatten()
+                    .collect::<HashSet<_>>();
+
+                // Remove components that have gone away.
+                sinks.retain(|name, _| {
+                    outputs.contains_key(name) || {
+                        debug!(message = "Removing component.", component_name = ?name);
+                        false
+                    }
+                });
+
+                // Keep the outputs for later clean-up when a shutdown is triggered.
+                last_outputs = Some(outputs);
+
+                // Send notifications to the client. The # of notifications will always be
+                // exactly equal to the number of patterns, so we can pre-allocate capacity.
+                let mut notifications = Vec::with_capacity(patterns.len());
+
+                // Matched notifications.
+                for pattern in matched.difference(&last_matches) {
+                    notifications.push(send_matched(tx.clone(), pattern).boxed());
+                }
+
+                // Not matched notifications.
+                for pattern in HashSet::from_iter(&patterns).difference(&matched) {
+                    notifications.push(send_not_matched(tx.clone(), pattern).boxed());
+                }
+
+                // Send all events. If any event returns an error, this means the client
+                // channel has gone away, so we can break out the loop.
+                if try_join_all(notifications).await.is_err() {
+                    debug!("Couldn't send notification(s); tap gone away.");
+                    break;
+                }
+            }
+        }
+    }
+
+    // At this point, the tap handler is being shut down due to the client/subscription
+    // going away. Clean up tap sinks by disconnecting them from the components being observed.
+    if let Some(outputs) = last_outputs {
+        for (name, id) in sinks {
+            if let Some(control_tx) = outputs.get(&name) {
+                let _ = control_tx.send(fanout::ControlMessage::Remove(id));
```

Comment:
Another place where I'm wondering if we should be handling the result. Also, should we `await` here?

Reply:
A message could be printed to stderr for sure. By this point, a shutdown in the subscription has been triggered, so we're typically beyond reporting this to the client.

The fanout control channel is a synchronous `tokio::sync::mpsc`, so no `await` is needed in this case.

Reply:
> The fanout control channel is a synchronous `tokio::sync::mpsc`, so no `await` is needed in this case.

Aha 👍 

> A message could be printed to stderr for sure. By this point, a shutdown in the subscription has been triggered, so we're typically beyond reporting this to the client.

Makes sense. Could we just log then?

Reply:
Logged in 19aa7cd7e1bd9ee4a91bd8140a5e1448e9f8a7b1.